climakitae/selectors.py

```diff
@@ -334,38 +424,114 @@ class DataSelector(param.Parameterized):
     def _update_scenarios(self):
+        """
+        The scenarios available will depend on the resolution (more will be available for 9km
+        than 3km for WRF eventually). Also ensures that "Historical Climate" is not
+        redundantly displayed when "Append historical" is also selected.
+        """
         _list_of_scenarios = list(self._choices._scenarios[self.resolution].keys())
-        if self.append_historical:
-            if "Historical Climate" in _list_of_scenarios:
-                _list_of_scenarios.remove("Historical Climate")
         self.param["scenario"].objects = _list_of_scenarios
-        self.scenario = [
-            None
-        ]  # resetting this avoids indexing errors with prior values
+        if self.append_historical and self.scenario is not None:
+            if "Historical Climate" in self.scenario:
+                _scenarios = self.scenario
+                _scenarios.remove("Historical Climate")
+                self.scenario = _scenarios
+
+    @param.depends("scenario", "append_historical", watch=True)
+    def _update_time_slice_range(self):
+        """
+        Will discourage the user from selecting a time slice that does not exist for any
+        of the selected scenarios, by updating the default range of years.
+        """
+        low_bound, upper_bound = self.time_slice
+        if "Historical Reconstruction" not in self.scenario:
+            low_bound = 1980
+            if ("Historical Climate" not in self.scenario) and (self.append_historical == False):
+                low_bound = 2015
+            else:
+                low_bound = 1980
+        elif low_bound >= 1980:
+            low_bound = 1950
+        if not True in ["SSP" in one for one in self.scenario]:
+            if "Historical Reconstruction" in self.scenario:
+                upper_bound = 2022
+            else:
+                upper_bound = 2015
+        elif upper_bound <= 2022:
+            upper_bound = 2100
+
+        self.time_slice = (low_bound, upper_bound)
 
     area_average = param.Boolean(default=False)
 
+    @param.depends("time_slice", "scenario", "append_historical", watch=False)
+    def view(self):
+        """
+        Displays a timeline to help the user visualize the time ranges available,
+        and the subset of time slice selected.
+        """
+        fig0 = Figure(figsize=(3, 2))
+        ax = fig0.add_subplot(111)
+        ax.spines["right"].set_color("none")
+        ax.spines["left"].set_color("none")
+        ax.yaxis.set_major_locator(ticker.NullLocator())
+        ax.spines["top"].set_color("none")
+        ax.xaxis.set_ticks_position("bottom")
+        ax.set_xlim(1950, 2100)
+        ax.set_ylim(0, 1)
+        ax.xaxis.set_major_locator(ticker.AutoLocator())
+        ax.xaxis.set_minor_locator(ticker.AutoMinorLocator())
+        mpl_pane = pn.pane.Matplotlib(fig0, dpi=144)
+
+        def update_bars(scenario, y_offset):
```

Comment:
Same comment about closures applies here, this can be inlined or if it's intended to be used elsewhere, as a method or module level function.

Reply:
This time, though, it is called multiple times, once for each scenario.